### PR TITLE
3523 hosts pools ScheduledAgentUpdates issue

### PR DESCRIPTION
**Pull request by ThojoUno**
# Description
Current desktop-virtualization/host-pools/main.bicep module has an incomplete agentUpdate object, which causes a 404 error when selecting Scheduled agent updates in the Settings pane for the Azure Virtual Desktop host pool.

Updated main.bicep host pools module adds parameters:
- agentUpdateType
- agentUpdateMaintenanceWindowHour
- agentUpdateMaintenanceWindowDayOfWeek
- agentUpdateMaintenanceWindows
- agentUpdateMaintenanceWindowTimeZone
- agentUpdateuseSessionHostLocalTime

These parameters are used to populate the agentUpdate object, which resolves the 404 error.

> Please also include the context.
> List any dependencies that are required for this change.

## Pipeline references
No pipeline updates required

| Pipeline |
| - |
| |

# Type of Change

Please delete options that are not relevant.

- [X] Bugfix (non-breaking change which fixes an issue)

# Checklist

- [X] I'm sure there are no other open Pull Requests for the same update/change
- [X] My corresponding pipelines / checks run clean and green without any errors or warnings
- [X] My code follows the style guidelines of this project
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation (readme)
- [X] I did format my code

### CODE REVIEW

**Comment by ThojoUno** [comment]
@microsoft-github-policy-service agree [company="{lunavi}"]

**Comment by ThojoUno** [comment]
@microsoft-github-policy-service agree company="lunavi"

File: modules/desktop-virtualization/host-pools/main.bicep [hunk 1]
```diff
@@ -101,11 +101,56 @@ param startVMOnConnect bool = false
...
+  'Friday'
+  'Saturday'
+])
+param agentUpdateMaintenanceWindowDayOfWeek string = 'Sunday'
+
+@sys.description('Optional. List of maintenance windows for scheduled agent updates.')
+param agentUpdateMaintenanceWindows array = [
+  {
+    hour: agentUpdateMaintenanceWindowHour
+    dayOfWeek: agentUpdateMaintenanceWindowDayOfWeek
```
Comment (AlexanderSehr) [inline modules/desktop-virtualization/host-pools/main.bicep:133]: Wouldn't it make sense to have the defaults for this one default maintenance window here? I.e.:
```suggestion
    hour: 22
    dayOfWeek: 'Sunday'
```

As soon as somebody where to introduce a custom (set) of window(s), the other default parameter values woudn't be applied anymore.

The same could be set for the `agentUpdate` object in general. It's default value could be:

```PowerShell
param agentUpdate object = {
  type: agentUpdateType
  maintenanceWindows: [
    {
      hour: 22
      dayOfWeek: 'Sunday'
    }
  ]
  maintenanceWindowTimeZone: 'Central Standard Time'
  useSessionHostLocalTime: false
}
```

File: modules/desktop-virtualization/host-pools/main.bicep [hunk 2]
```diff
@@ -101,11 +101,56 @@ param startVMOnConnect bool = false
 @sys.description('Optional. Array of role assignment objects that contain the \'roleDefinitionIdOrName\' and \'principalIds\' to define RBAC role assignments on this resource. In the roleDefinitionIdOrName attribute, you can provide either the display name of the role definition, or its fully qualified ID in the following format: \'/providers/Microsoft.Authorization/roleDefinitions/c2f4ef07-c644-48eb-af81-4b1b4947fb11\'.')
 param roleAssignments array = []
 
+// begin issue #3523 parameters fix section
```
Comment (AlexanderSehr) [inline modules/desktop-virtualization/host-pools/main.bicep:104]:
```suggestion
```

File: modules/desktop-virtualization/host-pools/main.bicep [hunk 3]
```diff
@@ -101,11 +101,56 @@ param startVMOnConnect bool = false
...
+  }
+]
+
+@sys.description('Optional. Time zone for scheduled agent updates.')
+param agentUpdateMaintenanceWindowTimeZone string = 'Central Standard Time'
+
+@sys.description('Optional. Whether to use localTime of the virtual machine for scheduled agent updates.')
+param agentUpdateUseSessionHostLocalTime bool = false
+
+// issue #3523 update existing agentUpdate object with required properties
```
Comment (AlexanderSehr) [inline modules/desktop-virtualization/host-pools/main.bicep:143]:
```suggestion
```

File: modules/desktop-virtualization/host-pools/main.bicep [hunk 4]
```diff
@@ -101,11 +101,56 @@ param startVMOnConnect bool = false
...
 @sys.description('Optional. The session host configuration for updating agent, monitoring agent, and stack component.')
 param agentUpdate object = {
-  useSessionHostLocalTime: true
+  type: agentUpdateType
+  maintenanceWindows: agentUpdateMaintenanceWindows
+  maintenanceWindowTimeZone: agentUpdateMaintenanceWindowTimeZone
+  useSessionHostLocalTime: agentUpdateUseSessionHostLocalTime
 }
 
+// end issue #3523 parameters fix section
```
Comment (AlexanderSehr) [inline modules/desktop-virtualization/host-pools/main.bicep:152]:
```suggestion
```

**Comment by AlexanderSehr** [comment]
Great PR @ThojoUno, just a few thoughts added as comments.

File: modules/desktop-virtualization/host-pools/main.bicep (hunk 1 above)
Comment (ThojoUno) [reply modules/desktop-virtualization/host-pools/main.bicep:133]: I've created parameters for those values, so you have the option to change them
```
@sys.description('Optional. Update hour for scheduled agent updates.')
@minValue(1)
@maxValue(23)
param agentUpdateMaintenanceWindowHour int = 22

@sys.description('Optional. Update day for scheduled agent updates.')
@allowed([
  'Sunday'
  'Monday'
  'Tuesday'
  'Wednesday'
  'Thursday'
  'Friday'
  'Saturday'
])
param agentUpdateMaintenanceWindowDayOfWeek string = 'Sunday'
```

**Review by ThojoUno — COMMENTED** [review]
suggested changes (removal of comments) committed to PR

File: modules/desktop-virtualization/host-pools/main.bicep (hunk 1 above)
Comment (AlexanderSehr) [reply modules/desktop-virtualization/host-pools/main.bicep:132]: Come to think of it a bit more, I think you're right. As the `agentUpdate` parameter is an object, this makes sense. My thinking got a bit stuck on that array I commented first.

**Review by AlexanderSehr — CHANGES_REQUESTED** [review]
Have you been able to test the changes in the pipeline using the test files we provide? We'd like to make sure that they will work once the change is merged into main.

If not, the next question would be whether you do have a test environment / subscription available to test in.

If not (again), I would suggest that I create an intermediate branch in CARML upstream, we merge your PR into there and I test it in our test environment before merging the intermediate branch ultimately to main. Your contribution would not be affected by this and it is simply a precaution.

**Comment by ThojoUno** [comment]
Yes, they pass all pipeline tests, and deploy via the test files. I've been running the new templates in production for several weeks now

**Review by AlexanderSehr — APPROVED** [review]

**Pull request merged**